pvanalytics/features/clipping.py

```diff
@@ -88,3 +88,115 @@ def levels(ac_power, window=4, fraction_in_window=0.75,
+
+
+def _clipping_power(ac_power, clip_derivative=0.0035, freq=None):
+    # Returns the clipping power threshold or None if no clipping is
+    # identified in the data.
+    #
+    # Copyright (c) 2020 Alliance for Sustainable Energy, LLC.
+    if not freq:
+        freq = pd.Timedelta(pd.infer_freq(ac_power.index)).seconds * 60
+    elif isinstance(freq, str):
+        freq = pd.Timedelta(freq).seconds * 60
+
+    # Use the derivative of the 99.5% quantile of daytime power at
+    # each minute to identify clipping.
+    powercurve = _daytime_powercurve(ac_power)
+    normalized_power = powercurve / powercurve.max()
+    power_derivative = (normalized_power.diff()
+                        / normalized_power.index.to_series().diff()) * freq
+    power_median = powercurve.median()
+
+    powersum = 0
+    count = 0
+    longest_powersum = 0
+    longest_count = 0
+    for derivative, power in zip(power_derivative, powercurve):
+        if _clipped(power, derivative, clip_derivative, power_median * 0.75):
+            count += 1
+            powersum += power
+        else:
+            count = 0
+            powersum = 0
+
+        if count > longest_count:
+            longest_count = count
+            longest_powersum = powersum
+
+    if (longest_count * freq) >= 60:
+        return longest_powersum / longest_count
+
+    return None
+
+
+def threshold(ac_power, clip_derivative=0.0035, freq=None):
+    """Detect clipping based on a maximum power threshold.
+
+    A clipping threshold is calculated from the AC power data and any
+    power greater than the threshold is flagged as clipped.
+
+    To calculate the clipping threshold, `ac_power` is aggregated at
+    each minute of the day. Low power data is removed to eliminate
+    night-time periods and the 99.5% quantile is computed. If the
+    derivative of the 99.5% quantile is less than `clip_derivative`
+    for a continuous period of at least one hour then clipping is
+    indicated. The mean power for that period is used as the
+    threshold. If there are multiple periods with a derivative less
+    than `clip_derivative` then the longest period is used to compute
+    the threshold.
+
+    Parameters
+    ----------
+    ac_power : Series
+        DatetimeIndexed series of AC power data.
+    clip_derivative : float, default 0.0035
+        Minimum derivative for clipping to be indicated. The default
+        value has been derived empirically to prevent false positives
+        for tracking PV systems.
+    freq : string, default None
+        A pandas string offset giving the frequency of data in
+        `ac_power`. If None then the frequency is infered from the
+        series index.
+
+    Returns
+    -------
+    Series
+        True when clipping is indicated.
+
+    Notes
+    -----
+    This function is based on code from the pvfleets_qa_analysis
+    project. Copyright (c) 2020 Alliance for Sustainable Energy, LLC.
```

Comment:
Here, I'm now unsure about the copyright. The algorithm is from pvfleets but the code is original here.

Reply:
I'm also unsure. The code here continues to diverge, so maybe we remove the copyright notice and just credit that it is based on pvfleest_qa_analysis?

Reply:
Yes, where the method is the same but the implementation is substantially different. Where the pvfleets code is reused (with minor changes such as variable names) we should carry their copyright.